compiler/passes/src/type_checking/checker.rs

```diff
@@ -933,127 +936,110 @@ impl<'a, N: Network> TypeChecker<'a, N> {
     pub(crate) fn check_function_signature(&mut self, function: &Function) {
         self.scope_state.variant = Some(function.variant);
 
-        // Special type checking for finalize blocks. Can skip for stubs.
+        let mut inferred_inputs: Vec<Type> = Vec::new();
+
         if self.scope_state.variant == Some(Variant::AsyncFunction) && !self.scope_state.is_stub {
-            // Finalize functions are not allowed to return values.
+            // Async functions are not allowed to return values.
             if !function.output.is_empty() {
                 self.emit_err(TypeCheckerError::async_function_cannot_return_value(function.span()));
             }
 
-            // Check that the input types are consistent with when the function is invoked.
-            if let Some(inferred_input_types) = self.async_function_input_types.get(&self.scope_state.location()) {
-                // Check same number of inputs as expected.
-                if inferred_input_types.len() != function.input.len() {
-                    return self.emit_err(TypeCheckerError::async_function_input_length_mismatch(
-                        inferred_input_types.len(),
-                        function.input.len(),
-                        function.span(),
-                    ));
-                }
-                // Check that the input parameters match the inferred types from when the async function is invoked.
-                for (t1, t2) in function.input.iter().zip_eq(inferred_input_types.iter()) {
-                    // Allow partial type matching of futures since inferred are fully typed, whereas AST has default futures.
-                    if matches!(t2, Type::Future(_)) && matches!(t1.type_(), Type::Future(_)) {
-                        // Insert to symbol table
-                        if let Err(err) = self.symbol_table.borrow_mut().insert_variable(
-                            Location::new(None, t1.identifier().name),
-                            self.scope_state.program_name,
-                            VariableSymbol {
-                                type_: t2.clone(),
-                                span: t1.identifier.span(),
-                                declaration: VariableType::Input(Mode::Public),
-                            },
-                        ) {
-                            self.handler.emit_err(err);
-                        }
-                    }
-                    self.check_eq_types(&Some(t1.type_().clone()), &Some(t2.clone()), t1.span())
-                }
+            let st = self.symbol_table.borrow();
+            let mut caller_finalizers = self
```

Comment:
Nit. A comment on what `caller_finalizers` is could be helpful to future readers.